# Data analysis: the average number of "hazardous" and "immediately hazardous" housing maintenance code violations recorded per building across NYC from Jan. 1, 2018 to Dec. 31, 2022

Data was downloaded from [NYC Open Data](https://data.cityofnewyork.us/Housing-Development/Housing-Maintenance-Code-Violations/wvxf-dwi5/explore/query/SELECT%0A%20%20%60violationid%60%2C%0A%20%20%60buildingid%60%2C%0A%20%20%60registrationid%60%2C%0A%20%20%60boroid%60%2C%0A%20%20%60boro%60%2C%0A%20%20%60housenumber%60%2C%0A%20%20%60lowhousenumber%60%2C%0A%20%20%60highhousenumber%60%2C%0A%20%20%60streetname%60%2C%0A%20%20%60streetcode%60%2C%0A%20%20%60zip%60%2C%0A%20%20%60apartment%60%2C%0A%20%20%60story%60%2C%0A%20%20%60block%60%2C%0A%20%20%60lot%60%2C%0A%20%20%60class%60%2C%0A%20%20%60inspectiondate%60%2C%0A%20%20%60approveddate%60%2C%0A%20%20%60originalcertifybydate%60%2C%0A%20%20%60originalcorrectbydate%60%2C%0A%20%20%60newcertifybydate%60%2C%0A%20%20%60newcorrectbydate%60%2C%0A%20%20%60certifieddate%60%2C%0A%20%20%60ordernumber%60%2C%0A%20%20%60novid%60%2C%0A%20%20%60novdescription%60%2C%0A%20%20%60novissueddate%60%2C%0A%20%20%60currentstatusid%60%2C%0A%20%20%60currentstatus%60%2C%0A%20%20%60currentstatusdate%60%2C%0A%20%20%60novtype%60%2C%0A%20%20%60violationstatus%60%2C%0A%20%20%60rentimpairing%60%2C%0A%20%20%60latitude%60%2C%0A%20%20%60longitude%60%2C%0A%20%20%60communityboard%60%2C%0A%20%20%60councildistrict%60%2C%0A%20%20%60censustract%60%2C%0A%20%20%60bin%60%2C%0A%20%20%60bbl%60%2C%0A%20%20%60nta%60/page/filter) as a .csv file

The data was filtered for selected years and for just "Class B and C" violations, on the NYC Open Data portal. Class B and C violations indicate conditions considered to be hazardous to the health and safety of tenants [(NYC Housing Preservation and Development)](https://www.nyc.gov/site/hpd/services-and-information/penalties-and-fees.page) 

This analysis, completed using the GroupBy function in Python Pandas, will then be compared to the average number of hazardous and immediately hazardous violations in the 13 buildings owned by landlord Robert Izsak.  

In [2]:
import pandas as pd

In [3]:
dtypes = {
    "ViolationID":"str",
    "BuildingID":"str"
    
}


housing_data = pd.read_csv(
    "../data/Housing_Maintenance_Code_Violations-6.csv",
      dtype=dtypes
)

/var/folders/ts/n_0bnjfs37vfxny5hw2lr6_80000gn/T/ipykernel_75296/1246859077.py:8: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  housing_data = pd.read_csv(


In [4]:
housing_data

ViolationID BuildingID  RegistrationID  BoroID        Borough  \
0         13957715     146093          348239       3       BROOKLYN   
1         13975317     119085          211560       2          BRONX   
2         14018854     744424          500617       5  STATEN ISLAND   
3         13998721      43514          105846       1      MANHATTAN   
4         13998722      43514          105846       1      MANHATTAN   
...            ...        ...             ...     ...            ...   
784363    15611001     939397          431260       4         QUEENS   
784364    15689756      50311          212324       2          BRONX   
784365    16027755     932377          226443       2          BRONX   
784366    16271443      20629          140066       1      MANHATTAN   
784367    16271444     376394          302714       3       BROOKLYN   

       HouseNumber LowHouseNumber HighHouseNumber       StreetName  \
0              225            225             225        23 STREET   
1             1890           1890            1890    WALTON AVENUE   
2              192            192             192  HEBERTON AVENUE   
3              552            552             554  WEST 185 STREET   
4              552            552             554  WEST 185 STREET   
...            ...            ...             ...              ...   
784363      148-15         148-13          148-19        90 AVENUE   
784364         289            289             289     BONNER PLACE   
784365         507            507             507  EAST 161 STREET   
784366         107            107             131  EAST 126 STREET   
784367        1373           1373            1373   STERLING PLACE   

        StreetCode  ... ViolationStatus RentImpairing   Latitude  Longitude  \
0             7480  ...           Close             N  40.660628 -73.995235   
1            72720  ...           Close             N  40.849519 -73.909293   
2            32450  ...            Open             N  40.635605 -74.132229   
3            37450  ...           Close             N  40.851277 -73.931018   
4            37450  ...           Close             N  40.851277 -73.931018   
...            ...  ...             ...           ...        ...        ...   
784363       18140  ...           Close             N  40.703307 -73.806940   
784364       13520  ...            Open             N  40.827652 -73.916712   
784365       26490  ...            Open             N  40.823066 -73.910004   
784366       19470  ...            Open             N  40.805507 -73.937997   
784367       81430  ...            Open             N  40.671593 -73.932887   

        CommunityBoard CouncilDistrict CensusTract        BIN           BBL  \
0                  7.0            38.0       145.0  3009467.0  3.006460e+09   
1                  5.0            14.0     23301.0  2007904.0  2.028270e+09   
2                  1.0            49.0       207.0  5023964.0  5.010210e+09   
3                 12.0            10.0       269.0  1063766.0  1.021570e+09   
4                 12.0            10.0       269.0  1063766.0  1.021570e+09   
...                ...             ...         ...        ...           ...   
784363            12.0            24.0       240.0        NaN           NaN   
784364             4.0            16.0       173.0  2001993.0  2.024230e+09   
784365             3.0            17.0       141.0  2116648.0  2.023660e+09   
784366            11.0             9.0       242.0  1054496.0  1.017750e+09   
784367             8.0            36.0       345.0  3036533.0  3.013720e+09   

                                NTA  
0                  Sunset Park West  
1                        Mount Hope  
2                     Port Richmond  
3          Washington Heights North  
4          Washington Heights North  
...                             ...  
784363                      Jamaica  
784364  Concourse-Concourse Village  
784365                   Morrisania  
784366          East Ha

In [5]:
housing_data.columns

Index(['ViolationID', 'BuildingID', 'RegistrationID', 'BoroID', 'Borough',
       'HouseNumber', 'LowHouseNumber', 'HighHouseNumber', 'StreetName',
       'StreetCode', 'Postcode', 'Apartment', 'Story', 'Block', 'Lot', 'Class',
       'InspectionDate', 'ApprovedDate', 'OriginalCertifyByDate',
       'OriginalCorrectByDate', 'NewCertifyByDate', 'NewCorrectByDate',
       'CertifiedDate', 'OrderNumber', 'NOVID', 'NOVDescription',
       'NOVIssuedDate', 'CurrentStatusID', 'CurrentStatus',
       'CurrentStatusDate', 'NovType', 'ViolationStatus', 'RentImpairing',
       'Latitude', 'Longitude', 'CommunityBoard', 'CouncilDistrict',
       'CensusTract', 'BIN', 'BBL', 'NTA'],
      dtype='object')

# Used the function .groupby to determine the count of ViolationID (a unique value for each violation) associated with each BuildingID (the same ID is assigned to each building)

# Used .describe() to view the average count of ViolationIDs associated with each BuildingID

In [6]:
violation_tallies = housing_data.groupby(
    ["BuildingID"]
    )["ViolationID"].count(
   
    ).reset_index(
).sort_values(
        by="ViolationID",
        ascending = False
)

violation_tallies

BuildingID  ViolationID
1606      111945          647
54234      94506          574
22259     344337          566
16240      28946          546
40059      64592          519
...          ...          ...
42115     676760            1
15972     287586            1
15959     287571            1
15954     287561            1
14687      27367            1

[57188 rows x 2 columns]

# According to the analysis the cumulative average number of class B and C HPD violations in NYC apartment buildings between 2018 and 2022 was 13.7

## However, there were large outliers - the maximum number of violations in a single building was 647, and 50% of buildings contained 4 or fewer violations. 

In [7]:
violation_tallies.describe()

ViolationID
count  57188.000000
mean      13.715605
std       29.169384
min        1.000000
25%        2.000000
50%        4.000000
75%       12.000000
max      647.000000

## To compare the average number of class B and C violations per building in the city with the buildings owned by landlord Robert Izsak, we downloaded a .csv from [NYC Open Data](https://data.cityofnewyork.us/Housing-Development/Housing-Maintenance-Code-Violations/wvxf-dwi5/explore/query/SELECT%0A%20%20%60violationid%60%2C%0A%20%20%60buildingid%60%2C%0A%20%20%60registrationid%60%2C%0A%20%20%60boroid%60%2C%0A%20%20%60boro%60%2C%0A%20%20%60housenumber%60%2C%0A%20%20%60lowhousenumber%60%2C%0A%20%20%60highhousenumber%60%2C%0A%20%20%60streetname%60%2C%0A%20%20%60streetcode%60%2C%0A%20%20%60zip%60%2C%0A%20%20%60apartment%60%2C%0A%20%20%60story%60%2C%0A%20%20%60block%60%2C%0A%20%20%60lot%60%2C%0A%20%20%60class%60%2C%0A%20%20%60inspectiondate%60%2C%0A%20%20%60approveddate%60%2C%0A%20%20%60originalcertifybydate%60%2C%0A%20%20%60originalcorrectbydate%60%2C%0A%20%20%60newcertifybydate%60%2C%0A%20%20%60newcorrectbydate%60%2C%0A%20%20%60certifieddate%60%2C%0A%20%20%60ordernumber%60%2C%0A%20%20%60novid%60%2C%0A%20%20%60novdescription%60%2C%0A%20%20%60novissueddate%60%2C%0A%20%20%60currentstatusid%60%2C%0A%20%20%60currentstatus%60%2C%0A%20%20%60currentstatusdate%60%2C%0A%20%20%60novtype%60%2C%0A%20%20%60violationstatus%60%2C%0A%20%20%60rentimpairing%60%2C%0A%20%20%60latitude%60%2C%0A%20%20%60longitude%60%2C%0A%20%20%60communityboard%60%2C%0A%20%20%60councildistrict%60%2C%0A%20%20%60censustract%60%2C%0A%20%20%60bin%60%2C%0A%20%20%60bbl%60%2C%0A%20%20%60nta%60/page/filter), filtered for the 13 BuildingIDs assigned to each building, for class B and C violations between Jan 1, 2018 and Dec 31, 2022. 

## Followed the same steps as listed above for the set of city-wide violations. 

In [8]:
notable_buildings = pd.read_csv("../data/notable_buildings.csv")

notable_buildings

ViolationID  BuildingID  RegistrationID  BoroID   Borough  HouseNumber  \
0       15183291       74521          202798       2     BRONX         1701   
1       15183292       74521          202798       2     BRONX         1701   
2       13737518       74521          202798       2     BRONX         1701   
3       12232648       74521          202798       2     BRONX         1701   
4       12388270       74521          202798       2     BRONX         1701   
..           ...         ...             ...     ...       ...          ...   
585     13302764      202731          374084       3  BROOKLYN          268   
586     13625809      202731          374084       3  BROOKLYN          268   
587     13625813      202731          374084       3  BROOKLYN          268   
588     13467116      202731          374084       3  BROOKLYN          268   
589     13467117      202731          374084       3  BROOKLYN          268   

     LowHouseNumber  HighHouseNumber       StreetName  StreetCode  ...  \
0              1701             1701  EASTBURN AVENUE       29720  ...   
1              1701             1701  EASTBURN AVENUE       29720  ...   
2              1701             1701  EASTBURN AVENUE       29720  ...   
3              1701             1701  EASTBURN AVENUE       29720  ...   
4              1701             1701  EASTBURN AVENUE       29720  ...   
..              ...              ...              ...         ...  ...   
585             268              288    BAY 38 STREET       17030  ...   
586             268              288    BAY 38 STREET       17030  ...   
587             268              288    BAY 38 STREET       17030  ...   
588             268              288    BAY 38 STREET       17030  ...   
589             268              288    BAY 38 STREET       17030  ...   

     ViolationStatus RentImpairing   Latitude  Longitude  CommunityBoard  \
0               Open             N  40.844336 -73.908519               4   
1               Open             N  40.844336 -73.908519               4   
2               Open             N  40.844336 -73.908519               4   
3              Close             N  40.844336 -73.908519               4   
4              Close             N  40.844336 -73.908519               4   
..               ...           ...        ...        ...             ...   
585            Close             N  40.592914 -73.993843              11   
586            Close             Y  40.592914 -73.993843              11   
587            Close             Y  40.592914 -73.993843              11   
588             Open             N  40.592914 -73.993843              11   
589             Open             N  40.592914 -73.993843              11   

    CouncilDistrict CensusTract      BIN         BBL               NTA  
0                15       22902  2007399  2027940049        Mount Hope  
1                15       22902  2007399  2027940049        Mount Hope  
2                15       22902  2007399  2027940049        Mount Hope  
3                15       22902  2007399  2027940049        Mount Hope  
4                15       22902  2007399  2027940049        Mount Hope  
..              ...         ...      ...         ...               ...  
585              43         304  3187601  3069280050  Bensonhurst East  
586              43         304  3187601  3069280050  Bensonhurst East  
587              43         304  3187601  3069280050  Bensonhurst East  
588              43         304  3187601  3069280050  Gravesend (West)  
589              43         304  3187601  3069280050  Gravesend (West)  

[590 rows x 41 columns]

In [9]:
notable_violation_tallies = notable_buildings.groupby(
    ["BuildingID"]
    )["ViolationID"].count(
   
    ).reset_index(
).sort_values(
        by="ViolationID",
        ascending = False
)

notable_violation_tallies

BuildingID  ViolationID
8       323187          123
0        74521           94
10      347340           90
1       142778           58
3       192024           56
4       202731           34
7       301247           33
12      391660           29
11      347632           18
6       256023           17
9       347311           17
2       187832           12
5       206597            9

# The 13 buildings owned by Robert Izsak recorded an average of 45 class B and C HPD violations each, between 2018 and 2022, compared to the citywide average of 13.7

In [10]:
notable_violation_tallies.describe()

BuildingID  ViolationID
count      13.000000    13.000000
mean   255452.538462    45.384615
std     95660.048202    36.516066
min     74521.000000     9.000000
25%    192024.000000    17.000000
50%    256023.000000    33.000000
75%    347311.000000    58.000000
max    391660.000000   123.000000

In [11]:
violation_tallies.to_csv('../output/citywide_violation_tallies.csv')

In [12]:
notable_violation_tallies.to_csv('../output/izsak_building_violation_tallies.csv')